### ISSUE TITLE
Advanced cache: time-based

### ISSUE THREAD

**Issue by williamyeny** [williamyeny/duke-registration-enhancer#21]
Even though I just removed the caching solution using chrome.storage, it might be time to put it back in to make a more advanced cache. With a new caching system, we can make more deliberate calls to the API, meaning we can keep the number API calls of low while increasing the accuracy of data.

Current cache: on page load, it creates the cache as an empty object. As course data is retrieved, it is stored in the cache and subsequent identical calls grab data from the cache. The cache only clears when the page is reloaded.

## Problem
It doesn't take into account how recent or outdated the data is. As soon as the page is reloaded, all cached data is cleared, even if it was retrieved just seconds earlier. This means you could needlessly call the API again if you refresh quickly. This also means you could be on the site for literally days, and you'll still use days-old data because the page hasn't refreshed.

## Solution
A cache in which data expires after a certain time would be more optimal. Therefore, even if you refresh the page, the cache won't clear if the data is very recent. Furthermore, any outdated data will be deleted on a subsequent call if it is discovered it was pulled too long ago. This time-based cache could be implemented by adding a timestamp property in each cache item and comparing the timestamp to the current time.

## Expiration vs eviction
The reason I decided to go with expiration instead of eviction (e.g. LRU) is that eviction does not properly update consistently. For example, a person who repeatedly checks 100 courses will be constantly evicting the old data, even if it is perfectly recent. On the flipside, a person who repeatedly checks only 2 courses may not have their cache data evicted at all! Therefore, eviction is not optimal.

## The possible ultimate cache
Even with this fancy expiration, each client will need to have their own cache. What this means is that as the number of people who use this extension increases, the amount of calls to the API will undoubtedly increase as well no matter how efficient your caching solution is.

But imagine this: one cache for ALL the clients. Here's my vision: when a client (student) loads the API for the first time, it sends the data to be cached to a dedicated server and the server holds onto that data. So when a different client does the exact same API call, it first checks with the server, and if the server has that data cached, it will immediately return it with no call to the API.

This makes things super efficient. For example, if there are 100 million students grabbing the data of a single course, *we would only make 1 call to the API* since they share a cache. Without this system, the minimum amount would be 100 million calls because they have their own individual cache. 

However, there are some problems. The obvious one is that I would need to create a backend rather than just the extension. This would be yet another thing to develop and maintain. However, I'm thinking of using AWS Lambda to make things easier -- hence the unused "lambda" folder. 

The second problem is less obvious: security. I'm sure all Duke students are saints, but there's nothing stopping some smartass from sending wrong data into the shared cache and ruining it for everyone. In order to mitigate that I'd need to do the API calls in the backend, but the API requires Duke authentication. This means that all calls would be from the same account (probably mine) rather than the account of the person who is using the extension. A friend of mine suggested sending the user's cookies to the server in order to authenticate as them, but I feel like impersonating a student's account to perform an API call is illegal in some way.

In conclusion, it's a good idea. But it's not there yet.

**Comment by williamyeny** [comment]
Implemented. Uses chrome.sync to create a "cloud cache", which persists after reloading a page and updates the local cache. A simple timestamp comparison determines if any information is out of date. Currently, information is set to expire after 10 minutes.
